src/main/kotlin/mate/academy/ArrayList.kt

```diff
@@ -1,44 +1,155 @@
+        return "Index: $index, Size: $size"
+    }
+
+
+    private fun grow(): Array<T?> {
+        val oldCapacity = array.size
+        val multiplicator = if (oldCapacity == size) 1.5 else 1.0
+        val newCapacity = (oldCapacity * multiplicator + 1).toInt()
+        array = if (oldCapacity > 0 || array !== empty_arrayData) {
+            return Arrays.copyOf(array, newCapacity)
+        } else {
+            return arrayOfNulls<Any>(defCapacity) as Array<T?>
+        }
+        return array
+    }
+
+    companion object {
+        fun listToArray(list: List<*>): Array<Any?>? {
+            if (!list.isEmpty()) {
+                return Array(list.size()) { index -> list[index] }
+            }
+            return null
+        }
+    }
+
     override fun add(value: T?) {
-        // implementation
+        val index = cursor
+        this.add(value, index)
+        cursor = index + 1
     }
 
+
     override fun add(value: T?, index: Int) {
-        // implementation
+        if (index > size || index < 0) {
+            throw ArrayListIndexOutOfBoundsException(errorMsg(index))
+        } else {
+            val s = size;
+            if (s == array.size) {
+                array = grow()
+            }
+            System.arraycopy(array, index, array, index + 1, size - index)
+            array[index] = value
+            size = s + 1
+        }
     }
 
+
     override fun addAll(list: List<T>) {
-        // implementation
+        if (!list.isEmpty()) {
+                val tempArray = listToArray(list) as Array<T?>
+            val newCapacity = size + tempArray.size
+            array = Arrays.copyOf(array, newCapacity)
+            System.arraycopy(tempArray, 0, array, size, tempArray.size)
+            size = newCapacity
+        }
     }
 
     override fun get(index: Int): T {
-        // implementation
-        return null as T // Placeholder, replace with actual implementation
+        if (index >= size || index < 0) {
+            throw ArrayListIndexOutOfBoundsException(errorMsg(index))
+        } else {
+            return array[index] as T
+        }
     }
 
     override fun set(value: T, index: Int) {
-        // implementation
+        if (index >= size || index < 0) {
+            throw ArrayListIndexOutOfBoundsException(errorMsg(index))
+        } else {
+            array[index] = value
+        }
     }
 
     override fun remove(index: Int): T {
-        // implementation
-        return null as T // Placeholder, replace with actual implementation
+        if (index >= size || index < 0) {
+            throw ArrayListIndexOutOfBoundsException(errorMsg(index))
+        } else {
+            val newSize = size - 1
+            val oldValue = get(index)
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index)
+            }
+            size = newSize
+            array[newSize] = null
+            return oldValue
+        }
     }
 
```

Comment:
Remove can be more shorter if u replace
"if (array[index] == null)"
and "if (element.equals(array[index]))"
With new function, which will check equality for both null and normal value